### PR TITLE
Added new file upload functionality (close #9)

### LINKED ISSUE #9: Add capability to download/upload file from/to the server

### PR DESCRIPTION
**Pull request by vvuri**
- Added display sorting - first directories, then files
- Added file link download
- Added form for selecting and uploading a file
- Files are uploaded to the current subdirectory
- Set file size limit to 10MB
- Added renaming of files if a file with the same name exists
- A directory for loading testing is created automatically
- The directory for testing functionality has been moved to '/tests'

### CODE REVIEW

File: src/fs_server.js [hunk 1]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
+            limits:  { fileSize: this.maxUploadSize }
+        }).single('fileToUpload');
+
+        this.app.post('/', (req, res) => {
+            this.upload( req, res, err => {
+                if (err) {
+                    debug(`${req.file} ${err.code}`, '_updateError');
+                    switch (err.code) {
+                        case 'LIMIT_FILE_SIZE':
+                            res.end('Choosen file size is greater than ' + this.maxUploadSize);
```
Comment (helen-dikareva) [inline src/fs_server.js:59]:
```suggestion
                            res.end('Chosen file size is greater than ' + this.maxUploadSize);
```

File: src/fs_server.js [hunk 2]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
+                            break;
+                        case 'INVALID_FILE_TYPE':
+                            res.end('Choosen file is of invalid type');
+                            break;
+                        case 'ENOENT':
+                            res.end('Unable to store the file');
+                            break;
```
Comment (helen-dikareva) [inline src/fs_server.js:66]: Should not we have a default block for err with any type?

File: src/fs_server.js [hunk 3]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
+                    }
+                }
+                else
+                    res.redirect(req.get('Referer') || '/');
```
Comment (helen-dikareva) [inline src/fs_server.js:70]: Your `_upload` do the same thing as this line does.
And this line more about 'redirect' than about 'uploading'. Let's use the method in both cases, just rename it.

File: src/fs_server.js [hunk 4]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
+        this.currentDir = ``;
+        this.maxUploadSize = 10 * 1024 * 1024;
 
         debug(`ClassInit::     HOST: ${args.host}  PORT: ${args.port}  ROOT_DIR: ${args.rootDir}`, 'constructor');
         debug(`Environment::   HOST: ${process.env.HOST}  PORT: ${process.env.PORT}  ROOT_DIR: ${process.env.ROOT_DIR}`, 'constructor');
         debug(`Config.json::   HOST: ${hostname}  PORT: ${port}  ROOT_DIR: ${dirname}`, 'constructor');
         debug(`Export::        HOST: ${this.host}  PORT: ${this.port}  ROOT_DIR: ${this.rootDir}`, 'constructor');
     }
 
+    _configureUpload () {
```
Comment (helen-dikareva) [inline src/fs_server.js:32]: (minor)
I tnink we can include the `_configureUpload` call in the `_initApp` method.

File: tests/upload-test.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    before(async () => {
+        requester = createRequester();
+        testConfig.rootDir = 'tests/upload';
+        await createTestUploadDir([
+            `./${testConfig.rootDir}`,
+            `./${testConfig.rootDir}/subdir`,
+            `./${testConfig.rootDir}/subdir/subsubdir`
+        ]);
+        await clearDir(testConfig.rootDir, true);
```
Comment (helen-dikareva) [inline tests/upload-test.js:21]: I don't get it. Why do we need to remove kust created directory?

File: src/fs_server.js [hunk 6]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
         this.host = args.host || process.env.HOST || hostname;
         this.port = args.port || process.env.PORT || port;
         this.rootDir = args.rootDir || process.env.ROOT_DIR || dirname;
-        this.dirPaths = [];
+        this.dirPath = [];
+        this.currentDir = ``;
```
Comment (helen-dikareva) [inline src/fs_server.js:23]: Shouldn't `currentDir ` be initialized to `/`?

File: tests/upload-test.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    });
+
+    let runs = [
+        { it: 'root', path: '/' },
+        { it: 'subdir', path: '/subdir/' }
+    ];
+
+    runs.forEach(run => {
+        describe(`A chain of related tests. Upload file to ${run.it}`, () => {
```
Comment (helen-dikareva) [inline tests/upload-test.js:39]: We never use chains of related tests

File: tests/upload-test.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            it('The uploaded file line.png was written to disk', async () => {
+                const listFiles = await getDir(`${testConfig.rootDir}${run.path}`, 'utf-8');
+
+                expect(listFiles.some(file => {
+                    return file === 'line.png';
+                })).to.eql(true);
+            });
+
+            it('The uploaded file line.png was given in get request', async () => {
+                let isAvailabile = false;
```
Comment (helen-dikareva) [inline tests/upload-test.js:59]:
```suggestion
                let isAvailable = false;
```

File: tests/upload-test.js [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        await srv.stop();
+        await clearDir(testConfig.rootDir, true);
+    });
+
+    let runs = [
+        { it: 'root', path: '/' },
+        { it: 'subdir', path: '/subdir/' }
+    ];
+
+    runs.forEach(run => {
```
Comment (helen-dikareva) [inline tests/upload-test.js:38]: It's quite unreadable tests organization. When I collapse all blocks I can't see any descriptive information, only 
```js
runs = [
         ...
    ];

    runs.forEach(...)
```
It's better to leave `describe` blocks on the top level and use data-driven approach inside them.

E.g.
```js
describe('File format', () => {
    runs.forEach(run => {
        it(`File with ${run.it} extension should be uploaded`, () => {});
    });
});
```

File: src/fs_server.js [hunk 10]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
 
+    _configureUpload () {
+        this.storage = multer.diskStorage({
+            destination: (req, file, cb) => {
+                this.currentDir = req.body.savePath;
+                debug(`./${this.rootDir}${this.currentDir}`, '_configureUpload.destination');
+                cb(null, `./${this.rootDir}${this.currentDir}`);
+            },
+            filename: async (req, file, cb) => {
+                debug(file.originalname, '_configureUpload.filename');
```
Comment (helen-dikareva) [inline src/fs_server.js:40]: We can get `currentDir` from request and avoid keeping it on the top level of server

File: src/fs_server.js [hunk 11]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
         this.host = args.host || process.env.HOST || hostname;
         this.port = args.port || process.env.PORT || port;
         this.rootDir = args.rootDir || process.env.ROOT_DIR || dirname;
-        this.dirPaths = [];
+        this.dirPath = [];
+        this.currentDir = ``;
+        this.maxUploadSize = 10 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (helen-dikareva) [reply src/fs_server.js:24]: Or we can place in the `config.json`

File: package.json [hunk 12]
```diff
@@ -20,6 +20,7 @@
   "scripts": {
     "build": "babel src -d dist",
     "test": "npm run lint && mocha --require babel-register --require babel-polyfill --colors tests/*-test.js",
+    "debug-test": "npm run lint && mocha --require babel-register --require babel-polyfill --colors --inspect-brk tests/upload-test.js",
```
[the remark this answers is not in the document]
Comment (vvuri) [reply package.json:23]: Fix

File: src/fs_server.js [hunk 13]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
+        debug(listFiles.dirs, '_getHTMLDirList.dir');
         data += `<ul>`;
-        for (const item of listFiles)
-            data += `<li> ${item} (<A href="http://${this.host}:${this.port}${subdir}${item}">open</A>) </li>`;
+        for (const item of listFiles.dirs)
+            data += `<li> <b>${item}</b> (<A href="http://${this.host}:${this.port}${subdir}${item}">open</A>)</li>`;
+
+        debug(listFiles.files, '_getHTMLDirList.files');
+        for (const item of listFiles.files) {
+            data += `<li> ${item} (<A href="http://${this.host}:${this.port}${subdir}${item}">open</A>)` +
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Fix

File: src/fs_server.js [hunk 14]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
+        this.maxUploadSize = 10 * 1024 * 1024;
 
         debug(`ClassInit::     HOST: ${args.host}  PORT: ${args.port}  ROOT_DIR: ${args.rootDir}`, 'constructor');
         debug(`Environment::   HOST: ${process.env.HOST}  PORT: ${process.env.PORT}  ROOT_DIR: ${process.env.ROOT_DIR}`, 'constructor');
         debug(`Config.json::   HOST: ${hostname}  PORT: ${port}  ROOT_DIR: ${dirname}`, 'constructor');
         debug(`Export::        HOST: ${this.host}  PORT: ${this.port}  ROOT_DIR: ${this.rootDir}`, 'constructor');
     }
 
+    _configureUpload () {
+        this.storage = multer.diskStorage({
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Add this.app and  this.server to constructor, change storage and upload as local.

File: src/fs_server.js [hunk 15]
```diff
@@ -44,22 +124,24 @@ export default class StaticServer {
 
         debug(`Export::        HOST: ${this.host}  PORT: ${this.port}  ROOT_DIR: ${this.rootDir}`, '_resDirListFiles');
         debug(`Dir: ${subdir}  req url: ${req.url}`, '_resDirListFiles');
-        debug(`#getDir( ${this.rootDir} + ${subdir} = ${this.rootDir.concat(subdir)} )`, '_resDirListFiles');
+        debug(`#getDir( ${this.rootDir} + ${subdir} = ${this.rootDir.concat(subdir)}),  currentDir:${this.currentDir}`, '_resDirListFiles');
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Remove concat, add variables

File: src/fs_server.js [hunk 16]
```diff
@@ -44,22 +124,24 @@ export default class StaticServer {
...
 
         getDir(this.rootDir.concat(subdir), 'utf-8')
             .then(files => {
-                data = this._getHTMLDirList(subdir, files);
+                const listDirAndFiles = getListDirAndFiles(this.rootDir.concat(subdir), files);
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Add path.join in server and heper files

File: src/fs_server.js [hunk 17]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
     _getHTMLDirList (subdir, listFiles) {
+        debug(listFiles, '_getHTMLDirList');
         let data = `<h2>List Files in <i>${this.rootDir}${subdir}</i>:</h2>`;
 
+        debug(listFiles.dirs, '_getHTMLDirList.dir');
         data += `<ul>`;
-        for (const item of listFiles)
-            data += `<li> ${item} (<A href="http://${this.host}:${this.port}${subdir}${item}">open</A>) </li>`;
+        for (const item of listFiles.dirs)
+            data += `<li> <b>${item}</b> (<A href="http://${this.host}:${this.port}${subdir}${item}">open</A>)</li>`;
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Fix

File: src/fs_server.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: I planned to make this setting in config.json. 
Add as maxUploadFileSize

File: src/fs_server.js [hunk 18]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
+            });
+        }, this._upload);
+    }
+
+    _upload (req, res) {
+        debug(req.file.destination, 'POST.file.destination');
+        res.redirect(req.get('Referer') || '/');
+    }
+
+    async _getNewFileName (fileName) {
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Fixed

File: tests/helper.js [hunk 19]
```diff
@@ -38,6 +40,39 @@ export function parseLiList (text) {
...
+    listFileDir.files.map( file => {
+        fs.unlink(`${subdir}/${file}`, err => {
+            if (err)
+                console.log(`Error deleted file: ${file}, err.message`);
+        });
+    });
+
+    if (recursive) {
+        listFileDir.dirs.map( dir => {
+            clearDir(`${subdir}/${dir}`, recursive);
```
[the remark this answers is not in the document]
Comment (vvuri) [reply tests/helper.js:57]: Fixed

File: src/fs_server.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Modified getDir and getListSubDirectories.

File: tests/helper.js [hunk 20]
```diff
@@ -38,6 +40,39 @@ export function parseLiList (text) {
...
+    await Promise.all(listDir.map( async path => {
+        await fs.mkdir(path, () => {});
+    }));
+}
+
+export function getMD5sum (filePath) {
+    const md5File = require('md5-file');
+
+    md5File(filePath)
+        .then(hash => {
```
[the remark this answers is not in the document]
Comment (vvuri) [reply tests/helper.js:70]: Fix - md5File.sync

File: tests/helper.js [hunk 21]
```diff
@@ -38,6 +40,39 @@ export function parseLiList (text) {
...
+export async function createTestUploadDir (listDir) {
+    await Promise.all(listDir.map( async path => {
+        await fs.mkdir(path, () => {});
+    }));
+}
+
+export function getMD5sum (filePath) {
+    const md5File = require('md5-file');
+
+    md5File(filePath)
```
[the remark this answers is not in the document]
Comment (vvuri) [reply tests/helper.js:69]: Translated to synchronous function

File: src/fs_server.js [hunk 22]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
         this.host = args.host || process.env.HOST || hostname;
         this.port = args.port || process.env.PORT || port;
         this.rootDir = args.rootDir || process.env.ROOT_DIR || dirname;
-        this.dirPaths = [];
+        this.dirPath = [];
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Fix

File: src/fs_server.js (hunk 4 above)
Comment (vvuri) [reply src/fs_server.js]: Fix

File: tests/upload-test.js (hunk 5 above)
Comment (vvuri) [reply tests/upload-test.js:21]: An empty directory is not added to GIT. And in tests, I completely clear the contents of the test directory. Option with .gitkeep I do not like.

File: tests/upload-test.js (hunk 9 above)
Comment (vvuri) [reply tests/upload-test.js:38]: Fix

File: src/fs_server.js (hunk 10 above)
Comment (vvuri) [reply src/fs_server.js]: Fix

File: src/fs_server.js (hunk 6 above)
Comment (vvuri) [reply src/fs_server.js]: Changed dirPaths in _initApp as local variable.

File: tests/upload-test.js (hunk 7 above)
Comment (vvuri) [reply tests/upload-test.js]: Fix

File: src/fs_server.js (hunk 3 above)
Comment (vvuri) [reply src/fs_server.js]: Fix

File: src/fs_server.js [hunk 23]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
+
+                debug(newName, '_configureUpload.getNewName');
+                cb(null, newName);
+            }
+        });
+
+        this.upload = multer({
+            storage: this.storage,
+            limits:  { fileSize: this.maxUploadSize }
+        }).single('fileToUpload');
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: add constant htmlUploadTag

File: tests/class-method-test.js [hunk 24]
```diff
@@ -17,25 +17,25 @@ describe('StaticServer unit test for methods:', () => {
 
     describe('_getHTMLDirList()', () => {
         it('should return an empty list if called without arguments', () => {
-            const result = srv._getHTMLDirList('', '');
+            const result = srv._getHTMLDirList('', { dirs: '', files: '' });
```
[the remark this answers is not in the document]
Comment (vvuri) [reply tests/class-method-test.js:26]: Add test

File: tests/upload-test.js [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+        it(`Uploaded format file - ${run.it}`, async () => {
+            await requester
+                .post('/')
+                .field({ savePath: run.path })
+                .attach('fileToUpload', `${run.sourceDir}${run.fileName}`, run.fileName)
+                .then(result => {
+                    expect(result).to.have.status(200);
+                });
+
+            expect(getMD5sum(`${run.sourceDir}${run.fileName}`))
```
[the remark this answers is not in the document]
Comment (vvuri) [reply tests/upload-test.js:139]: Fix - to.equal

File: src/fs_server.js [hunk 26]
```diff
@@ -18,22 +19,101 @@ export default class StaticServer {
...
+        res.redirect(req.get('Referer') || '/');
+    }
+
+    async _getNewFileName (fileName) {
+        debug(fileName, '_getNewName');
+        debug(this.rootDir + this.currentDir, '_getNewName');
+        const listDirFiles = await getDir( this.rootDir + this.currentDir, 'utf-8' );
+
+        debug(listDirFiles, '_getNewName.listDirFiles');
+        const isFileNameMatch = listDirFiles.some( file => {
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Modified getNewFileName - copy file format name

File: src/fs_server.js (hunk 2 above)
Comment (vvuri) [reply src/fs_server.js]: Add default error message

File: package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (helen-dikareva) [reply package.json]: I asked you to remove linting. Did you forget or you are disagree with it?

[the remark this answers is not in the document]
Comment (vvuri) [reply package.json]: I agree, he is not needed here. Fixed.

File: src/fs_server.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (helen-dikareva) [reply src/fs_server.js]: I still see two the same long lines, we can 

- get rid of redundant replay
- improve readability

At leat we can extract `href` creation in a function, e.g. 
```js
const getItemHref = (subdir, item) => `"http://${this.host}:${this.port}${subdir}${item}"`;

for (const item of listFiles.dirs)
    data += `<li><A href=${getItemHref(subdir, item)}><b>${item}</b></A></li>`;

for (const item of listFiles.files) {
    const substr = `<A href=${getItemHref(subdir, item)}`;

    data += `<li>${substr}>${item}</A>` +
            ` (${substr} download>download</A>)</li>`;
}
```

File: tests/helper.js [hunk 27]
```diff
@@ -38,6 +40,35 @@ export function parseLiList (text) {
...
+export async function createTestUploadDir (listDir) {
+    await Promise.all(listDir.map( async dirPath => {
+        await fs.mkdir(dirPath, () => {});
+    }));
+}
+
+export function getMD5sum (filePath) {
+    const md5File = require('md5-file');
+
+    return md5File.sync(filePath);
```
Comment (helen-dikareva) [inline tests/helper.js:69]: We don't need this method. Move require in the top section of the test file and use `md5File.sync(filePath);` without this wrapper

File: tests/upload-test.js (hunk 5 above)
Comment (helen-dikareva) [reply tests/upload-test.js:21]: You didn't get my point right - https://github.com/helen-dikareva/node-express-stub/pull/16#discussion_r429768288
I said that in tests' `before` block you create folders and right after that try to clear upload folder. Why?
Can we skip clear, will it change something?

File: tests/helper.js [hunk 28]
```diff
@@ -38,6 +40,35 @@ export function parseLiList (text) {
         });
 }
 
+export async function clearDir (subdir, recursive = false) {
```
Comment (helen-dikareva) [inline tests/helper.js:43]: Use `path.join` here too

Comment (helen-dikareva) [reply tests/helper.js:43]: This method still returns control before all files and directories are deleted

File: src/fs_server.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Has become more readable.

File: tests/upload-test.js [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+
+    runs = [
+        { it: 'One symbol', fileName: '1', path: '/subdir/subsubdir/' },
+        { it: 'long', fileName: 'VeryLongNameOfFileDownloadOnServer.txt', path: '/' },
+        { it: 'Cyrillic', fileName: 'Название файла на русском языке.doc', path: '/subdir/subsubdir/' },
+        { it: 'Space', fileName: 'Test file 2 download.txt', path: '/subdir/subsubdir/' },
+        { it: 'Dots', fileName: 'File.with.dots', path: '/subdir/' }
+    ];
+
+    runs.forEach(run => {
```
Comment (helen-dikareva) [inline tests/upload-test.js:110]: Add describe block (see https://github.com/helen-dikareva/node-express-stub/pull/16#discussion_r429771901) for 
'Uploaded file names with ${run.it} name'
and 
'Uploaded format file - ${run.it}'
tests

File: src/fs_server.js [hunk 30]
```diff
@@ -1,85 +1,160 @@
...
+        </form>`;
+
         return data;
     }
 
     _resDirListFiles (req, res, next, subdir = '') {
         let data;
 
         subdir = req.url;
+        const currentPath = path.join(this.rootDir, subdir);
```
Comment (helen-dikareva) [inline src/fs_server.js:74]: It's better to use functions as 'pure' ones.
It means do not modify arguments:
`subdir = req.url;` -> `const localVariableName = req.url;`
In this case we'll be able to get rid of `..., next, subdir = ''`) arguments in the signature

File: src/fs_helper.js [hunk 31]
```diff
@@ -1,55 +1,85 @@
...
+            return result;
+        })
         .catch( err => {
             debug(err, 'getDir.error');
             throw err.message;
         });
 };
 
-const dirPaths = ['/'];
+export const getListSubDirectories = async (rootDir, subdir = '', dirPaths = ['/']) => {
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_helper.js]: Rename getListSubDirectories > getSubDirectoryUrlsRecursive

File: src/fs_helper.js [hunk 32]
```diff
@@ -1,55 +1,85 @@
...
             throw err.message;
         });
 
     return dirPaths;
 };
+
+export const getNewFileName = async (fileName, pathToFile, copy = 0) => {
+    const ext = path.extname(fileName);
+    const base = path.basename(fileName, ext);
+    let fileCandidate = `${base} (${copy})${ext}`;
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_helper.js:68]: it's clear

File: src/fs_server.js [hunk 33]
```diff
@@ -1,85 +1,159 @@
...
 import express from 'express';
+import multer from 'multer';
+import path from 'path';
 
-import { getDir, getListSubDirectories } from './fs_helper';
+import { getDir, getListSubDirectories, getNewFileName } from './fs_helper';
 import config from '../config.json';
 import debug from './fs_logger';
 
+const htmlUploadTag = 'fileToUpload';
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: it's clear

File: tests/helper.js (hunk 27 above)
Comment (vvuri) [reply tests/helper.js]: Fix

File: src/fs_server.js (hunk 30 above)
Comment (vvuri) [reply src/fs_server.js:74]: it's clear

File: tests/upload-test.js (hunk 29 above)
Comment (vvuri) [reply tests/upload-test.js:110]: Fix

File: tests/helper.js (hunk 28 above)
Comment (vvuri) [reply tests/helper.js:44]: Fix

File: src/fs_server.js [hunk 34]
```diff
@@ -1,85 +1,159 @@
...
-        for (const path of this.dirPaths) {
-            this.app.use(path, express.static(this.rootDir + path));
-            this.app.get(path, this._resDirListFiles.bind(this) );
+        debug(dirPaths, '_initApp');
+        for (const dirPath of dirPaths) {
+            this.app.use(dirPath, express.static(path.join(this.rootDir, dirPath)));
+            this.app.get(dirPath, this._resDirListFiles.bind(this) );
         }
+
+        const storage = multer.diskStorage({
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js:107]: it's clear

File: src/fs_server.js [hunk 35]
```diff
@@ -1,85 +1,159 @@
...
+                        case 'ENOENT':
+                            res.end('Unable to store the file');
+                            break;
+                        default:
+                            res.end('Something went wrong. Please try again later.');
+                            break;
+                    }
+                }
+                else
+                    res.redirect(req.get('Referer') || '/');
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_server.js]: Fix reloadPage method

File: src/fs_helper.js [hunk 36]
```diff
@@ -1,55 +1,85 @@
...
+
+    if ( copy === 0)
+        fileCandidate = fileName;
+
+    debug(`${pathToFile} : ${fileName} = ${base} + ${copy} + ${ext}`, 'getNewName');
+    try {
+        await fsStat( path.join(pathToFile, fileCandidate) );
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_helper.js:72]: Fix

File: src/fs_helper.js [hunk 37]
```diff
@@ -1,55 +1,85 @@
...
-            console.log(`Error getting information about a file:${fileName}: ${err.message}`);
-            return {
-                isDirectory: () => {
-                    return false;
-                },
-                error: new Error(`Error getting information about a file:${fileName}: ${err.message}`)
-            };
-        });
-};
+export const getDir = ( folder, enconding = 'utf-8' ) => {
```
[the remark this answers is not in the document]
Comment (vvuri) [reply src/fs_helper.js]: Fix

File: src/fs_helper.js [hunk 38]
```diff
@@ -1,55 +1,81 @@
...
         .catch( err => {
-            console.log(`Error read ${rootDir.concat(subdir)}: ${err}`);
+            console.log(`Error read ${currentPath}: ${err}`);
             throw err;
         });
 
-    await Promise.all( files.map( async fileName => {
-        const stat = await statDir(`${rootDir}${subdir}/${fileName}`);
+    await Promise.all( fileList.dirs.map( async fileName => {
+        debug( fileName, 'getListSubDirectories.map' );
```
Comment (helen-dikareva) [inline src/fs_helper.js:49]: You left the old method name `getListSubDirectories` in debug comments

File: src/fs_helper.js [hunk 39]
```diff
@@ -1,55 +1,81 @@
...
 };
+
+export const getNewFileName = async (fileName, pathToFile, copy = 0) => {
+    const ext = path.extname(fileName);
+    const base = path.basename(fileName, ext);
+    let fileCandidate = copy === 0 ? fileName : `${base} (${copy})${ext}`;
+
+    debug(`${pathToFile} : ${fileName} = ${base} + ${copy} + ${ext}`, 'getNewName');
+    try {
+        fs.lstatSync( path.join(pathToFile, fileCandidate) );
```
Comment (helen-dikareva) [inline src/fs_helper.js:71]: Why `getNewFileName ` is still async if you decided to use sync analog of `fsStat`?

File: src/fs_server.js [hunk 40]
```diff
@@ -1,85 +1,163 @@
...
+            data += `<li>${substr}>${item}</A>` +
+                ` (${substr} download>download</A>)</li>`;
+        }
         data += `</ul>`;
 
+        data += '<br>';
+        data += `
+        <form action="/" enctype="multipart/form-data" method="post">
+          <input type="hidden" id="savePath" name="savePath" value="${subdir}"">
+          <input type="file" name="fileToUpload" value="Select file">
```
Comment (helen-dikareva) [inline src/fs_server.js:58]: I've already asked you to reuse `HTML_UPLOAD_NAME ` here. Is it impossible or you have something against it?

File: src/fs_server.js [hunk 41]
```diff
@@ -1,85 +1,163 @@
...
+                debug(error, '_resDirListFiles.catch');
             });
     }
 
-    _initApp () {
+    _reloadPage (req, res) {
+        res.redirect(req.get('Referer') || '/');
+    }
+
+    async _configureUpload () {
```
Comment (helen-dikareva) [inline src/fs_server.js:96]: Why this method is async?

File: tests/upload-test.js [hunk 42]
```diff
@@ -0,0 +1,173 @@
...
+        { it: 'JPEG', sourceDir: './tests/public/elements/', fileName: 'zond.jpeg', path: '/subdir/' },
+        { it: 'PNG', sourceDir: './tests/public/elements/', fileName: 'line.png', path: '/subdir/subsubdir/' },
+        { it: 'SVG', sourceDir: './tests/public/elements/subelements/', fileName: 'logo2.svg', path: '/subdir/subsubdir/' },
+        { it: 'MP4', sourceDir: './tests/public/elements/', fileName: 'file_example_MP4_640_3MG.mp4', path: '/subdir/subsubdir/' }
+    ];
+
+    describe('Check Uploading file format', () => {
+        runs.forEach(run => {
+            it(`Uploaded format file - ${run.it}`, async () => {
+                const md5File = require('md5-file');
```
Comment (helen-dikareva) [inline tests/upload-test.js:142]: We always execute require in the top of the file. There is no problem with performance at all.

Comment (vvuri) [reply tests/upload-test.js]: Fix

File: src/fs_server.js (hunk 41 above)
Comment (vvuri) [reply src/fs_server.js]: Fix

File: src/fs_server.js (hunk 40 above)
Comment (vvuri) [reply src/fs_server.js]: Fix

File: src/fs_helper.js (hunk 39 above)
Comment (vvuri) [reply src/fs_helper.js:71]: Fix

File: src/fs_helper.js (hunk 38 above)
Comment (vvuri) [reply src/fs_helper.js]: Fix

File: tests/helper.js [hunk 43]
```diff
@@ -38,6 +41,30 @@ export function parseLiList (text) {
...
+    }
+}
+
+export async function createTestUploadDir (listDir) {
+    await Promise.all(listDir.map( async dirPath => {
+        await fs.mkdir(dirPath, () => {});
+    }));
+}
+
+export function deleteTestUploadDirs (rootUploadDir) {
```
Comment (helen-dikareva) [inline tests/helper.js:64]: This method is still just a wrapper the method call. We don't need it, we can use `fs.rmdirSync(rootUploadDir, { recursive: true });` directly

File: tests/helper.js [hunk 44]
```diff
@@ -38,6 +41,30 @@ export function parseLiList (text) {
         });
 }
 
+export async function clearDir (subdir, recursive = false) {
```
Comment (helen-dikareva) [inline tests/helper.js:44]: If you decided to give up using your `clearDir` method remove it.

**Pull request merged**